pkg/cmd/auth/status/status.go

```diff
@@ -220,6 +220,10 @@ func statusRun(opts *StatusOptions) error {
 		users := authCfg.UsersForHost(hostname)
 		for _, username := range users {
 			if username == activeUser {
+				_, ok := entry.(validEntry)
+				if err == nil && !ok {
+					err = cmdutil.SilentError
+				}
```

Comment:
Thanks for opening this PR to restore original `gh auth status` error exit code behavior, @Stausssi! 🤗

@williammartin : I'd appreciate your thoughts given your experience with multi-account support and whether my suggestions below make sense 🙇 

## History

I wanted to dig up on how this command worked before multi-account support was added so we offer a consistent experience:

`gh auth status` would build up information about the authentication state for each of the hostnames as 1 user per hostname was all that was supported:

https://github.com/cli/cli/blob/2f31607096b39b4954fb5a211166b2b9e627fcb6/pkg/cmd/auth/status/status.go#L105-L107

Unless there was no status information, `gh auth status` would print something to stdout or stderr and return silent err:

https://github.com/cli/cli/blob/2f31607096b39b4954fb5a211166b2b9e627fcb6/pkg/cmd/auth/status/status.go#L169-L196

## Alternate ideas

**1. Instead of checking within the `users` loop, does it make sense to test for immediately after building the entry before the loop?**
 
https://github.com/cli/cli/blob/f179342211468654f44d3a59bca77b8dc0ce8cb6/pkg/cmd/auth/status/status.go#L203-L218

**2. Do we need to restore the `stderr` vs `stdout` output logic around `failed`?**

https://github.com/cli/cli/blob/2f31607096b39b4954fb5a211166b2b9e627fcb6/pkg/cmd/auth/status/status.go#L169-L196

Reply:
Thanks for the extensive review, @andyfeller! I really appreciate the insight 💡 

I do like your suggestions and will get to work once we have the input from @williammartin
Still, I'd like to add my thoughts to your input.

## 1. Exit with 1 if any of the accounts are not logged in

**Tbh, I was and still am struggling with this** 🤔

In the end I've decided for `activeUser`, as that's arguably the crucial piece of information when invoking `gh auth status` inside a script. I suppose I'm slightly biased here, as that's what I was expecting in my script: I want to check via `gh auth status` if the following `git` and also `gh` commands are authenticated:

```bash
if ! gh auth status ; then
    # Login gh
fi

git pull
gh release download ...
```

In this case, exiting with 1 if an inactive account is not logged in, while the active account is in fact logged in, could also be seen as confusing to the user.

Looking at the [manual ](https://cli.github.com/manual/gh_auth_status) of `gh auth status`: 
> For each host, the authentication state of each known account is tested and any issues are included in the output. Each host section will indicate the active account, which will be used when targeting that host.

Note *active account, which will be used*.

---

Maybe this can be coupled with the `--hostname` flag? See [`gh auth setup-git`](https://cli.github.com/manual/gh_auth_setup-git):

> By default, GitHub CLI will be set as the credential helper for all authenticated hosts. If there is no authenticated hosts the command fails with an error.

=> If no `--hostname` **is given**, it fails if **any** account is not logged in

and

> Alternatively, use the `--hostname` flag to specify a single host to be configured. If the host is not authenticated with, the command fails with an error.

=> If `--hostname` is **not given**, if fails if the **active** account is not logged in.

---

Some commands, e.g. [`gh auth logout`](https://cli.github.com/manual/gh_auth_logout), [`gh auth switch`](https://cli.github.com/manual/gh_auth_switch) and [`gh auth token`](https://cli.github.com/manual/gh_auth_token) also provide a `--user` flag.
Should that be added for `gh auth status` as well and then be coupled to the exit code?

However, `gh auth token` states:

> Without the `--user` flag, the active account for the host is chosen.

---

**Independently of the final decision here, I'll document the behaviour in the manual of `gh auth status`**

## 2. Output to `stderr`

I like it :)
It will improve the handling of `gh auth status` inside scripts even further.

Just a small question: Do we only want the invalid entry in `stderr` and the valid ones in `stdout` or everything in `stderr` (only in the error case ofc)?

If I understand correctly, previous code only printed the invalid entry to `stderr`, which I would also propose.

---

Looking forward to your input!

Reply:
Hmmm lots of feelings about this. My understanding is that originally the intent behind this command was that it should be human readable and not for automation but then for whatever reason there was a decision to make some decisions around exit codes and [solidified a bit further later because people were actually trying to consume the output programmatically](https://github.com/cli/cli/issues/7447). 

Fundamentally, I don't think there is a good answer to the exit code issue for a plain old `gh auth status` command. I don't think the previous solution was particularly good either even before multi-account.

The issue is that "am I in a good state?" doesn't have sufficient enough context to provide a good answer via an exit code. For example, am I in a good state if my github.com active user token is valid but an inactive one isn't valid? We can begin to scope this down via `--hostname` and `--user` as suggested but it has one caveat. I think that the pattern implies a series of scope reductions:
  * No flags = any invalid user on any host is a failure
  * `--hostname` = any invalid user on that host is a failure
  * `--user` = any invalid user with that name on any host is a failure
  * `--hostname` and `--user` = a invalid specific user on a specific host is a failure

Above, it was said:

> => If --hostname is not given, if fails if the active account is not logged in.

But that precludes asking the question "are all the accounts on a host valid".

However, the caveat with this "scope reduction" approach is that it doesn't easily allow you to test the active user on a host (or across hosts) if that's what you really care about. So to resolve this we could introduce the idea of `--active` alongside `--user`. Thus, you could do `gh auth status --hostname github.com --active` to active the desired result in the previously mentioned script.

--

One downside to the approach above is that scripts need to be _specific_ from the outset. If you think `gh auth status` is going to work and later you log into another account or host, you may have surprising failures.

An alternative approach is that we could consider whether we should always treat `auth status` as success (except in the case of exceptional errors) and [provide programmatic access to the necessary information](https://github.com/cli/cli/issues/8637).

Reply:
Thanks for giving your perspective, @williammartin !

**Honestly speaking, I feel like this is something that has to be decided on your end.**

Judging by the issues you've linked I'm quite sure any change in behaviour of `gh auth status` should be considered breaking for a lot of scripts.

Keeping this in mind, I think combining the mentioned behaviour

> - No flags = any invalid user on any host is a failure
> - `--hostname` = any invalid user on that host is a failure
> - `--user` = any invalid user with that name on any host is a failure
> - `--hostname` and `--user` = a invalid specific user on a specific host is a failure

combined with the `--active` flag is the best way to move forward without breaking existing solutions.
The downside of having to be specific when invoking the command is acceptable imo and even guarantees a certain behaviour in the future.

What should be the expected behaviour if just the `--active` flag is passed without a hostname?

Finally, I'm generally a fan of adding the `--json` option. However, I'm struggling to see how it would simplify the approach of specifying flags. Ultimately, it doesn't matter whether I have to specify 2 flags (`--hostname` and `--user` or`--active`) or have a `jq` filter with exactly the same information for parsing the payload.

Reply:
### Suggestions

Thank you both for insights into how to unblock this! 🤗  I realize there is no single right solution on this one, so figuring out how to unblock this while giving users choice, so I'll throw out final thoughts based on your feedback:

1. `gh auth status` defaults to checking active accounts
2. `gh auth status --all` extends to checking active + inactive accounts
3. `gh auth status --user` is a follow up issue
4. `gh auth status` outputs to `stderr` when an invalid or problematic account is encountered under the above conditions

### Thoughts

> An alternative approach is that we could consider whether we should always treat auth status as success (except in the case of exceptional errors) and https://github.com/cli/cli/issues/8637.

This might be an option if people weren't using [`gh auth status` in workflows to test authentication](https://github.com/search?q=path%3A.github%2Fworkflows%2F*.yml%20%22gh%20auth%20status%22&type=code).

> In this case, exiting with 1 if an inactive account is not logged in, while the active account is in fact logged in, could also be seen as confusing to the user.

After looking at the automation cases linked above, **I'm starting to agree with you that maybe defaulting to the active accounts and having a flag that checks active + inactive / all is less impactful**.

Reply:
I'm fine with the suggestions, thank you!

Considering my proficiency with Go, which is basically non-existent, I'm glad to move `--user` into a separate issue. The `--all` flag as well as printing to `stderr` should be doable; I'll let you know if I need support there

Reply:
> This might be an option if people weren't using [gh auth status in workflows to test authentication](https://github.com/search?q=path%3A.github%2Fworkflows%2F*.yml%20%22gh%20auth%20status%22&type=code).

Mixed feelings about this. Since this was broken with the multi-account changes, reintroducing non-zero exit status may break anyone who started relying on that behaviour in the past 7 months. That said, with regards the exit codes, it's probably fair to say more people rely on the old behaviour (not knowing it is broken) than the new behaviour.

I can't see an obvious example of someone [depending on the previous `stderr` behaviour](https://github.com/search?q=%22gh+auth+status+%7C+%22&type=code) that would be broken be our change. On the flip side, I can see[ someone who would be broken by us changing it back](https://github.com/LongTermSupport/fedora-desktop/blob/4ebd9db2fc399b71fe7676037be5b9c54c36e9c7/run.bash#L147C27-L147C43):

```
if gh auth status | grep "Failed to log in"; then
```

If we reverted the `stdout/stderr` streams change, they would be broken. We can of course, inform them of this if we decide to change it back.

> I'm starting to agree with you that maybe defaulting to the active accounts and having a flag that checks active + inactive / all is less impactful. 

I think it's _kind of_ confusing if the exit codes are detached from the visual output. 

To resolve this we have two options:
 * Make the exit code non-zero if any account has authentication issues
 * Make the visual output only include active accounts by default, and use `--all` to bring them into view

For the latter, I think it would probably be a bit confusing for anyone that expects the current behaviour, and _technically_ breaking for anyone consuming the output of `auth status` in a machine readable way. I don't think this is a good idea but I wanted to present it.

> As for --user, I can see someone asking for that, however this is really beyond the scope of restoring the original behavior of exit codes.

That's a fair point. So to be clear, I think the options on the table for exit codes are:

**Given** any active account has issues with authentication
**When** I run `gh auth status`
**Then** the exit code is non-zero

**OR**

**Given** any (active or inactive) account has issues with authentication
**When** I run `gh auth status`
**Then** the exit code is non-zero

I think the second is preferable because it later allows us to scope things down with `--hostname` and `--user` without breaking people again, or introducing surprising UX. For example, if we do `--all` we need to explain that it only impacts the exit codes, and we avoid having to explain the combination of `--all` with other flags.

> Just a small question: Do we only want the invalid entry in stderr and the valid ones in stdout or everything in stderr (only in the error case ofc)?
>
> If I understand correctly, previous code only printed the invalid entry to stderr, which I would also propose.

The previous implementation printed all entries to `stderr` in the case of a failure. You can see that `failed` would be set once and used inside the loop for all entries. https://github.com/cli/cli/blob/2f31607096b39b4954fb5a211166b2b9e627fcb6/pkg/cmd/auth/status/status.go#L175C19-L175C25

I can see value in splitting the entries because it allows for grepping stdout for successful accounts. However, I do worry that it could be very, very, very confusing to debug. My gut also tells me that it's going to be really finicky to write and maintain the code that makes the output style correctly in both cases when it comes to newlines 😅 

I think if people really want the detail about a specific account we should encourage them to use filtering flags `--hostname` and `--user` or `--json` and `--jq`.

### Final Proposal

That was a lot of text, I'm sorry. In some ways I was gathering my own thoughts through the process of writing. My proposal is:

**Given** any (active or inactive) account has issues with authentication
**When** I run `gh auth status`
**Then** the exit code is non-zero
**And** all entries are printed to stderr

**Given** any (active or inactive) account on github.com has issues with authentication
**When** I run `gh auth status --hostname github.com`
**Then** the exit code is non-zero
**And** all entries are printed to stderr

**Given** any (active or inactive) account on github.com has issues with authentication
**When** I run `gh auth status --hostname ghe.io`
**Then** the exit code is zero
**And** all entries are printed to stdout

The last two are just distinguishing that `--hostname` should already apply the filtering.

I hope this makes sense 🙏

Reply:
> ### Final Proposal
> That was a lot of text, I'm sorry. In some ways I was gathering my own thoughts through the process of writing. My proposal is:
> 
> **Given** any (active or inactive) account has issues with authentication **When** I run `gh auth status` **Then** the exit code is non-zero **And** all entries are printed to stderr
> 
> **Given** any (active or inactive) account on github.com has issues with authentication **When** I run `gh auth status --hostname github.com` **Then** the exit code is non-zero **And** all entries are printed to stderr
> 
> **Given** any (active or inactive) account on github.com has issues with authentication **When** I run `gh auth status --hostname ghe.io` **Then** the exit code is zero **And** all entries are printed to stdout
> 
> The last two are just distinguishing that `--hostname` should already apply the filtering.
> 
> I hope this makes sense 🙏

Yes, and I agree 👍

Reply:
Again, thanks for your extensive review @andyfeller and @williammartin !

I was typing a long analysis of this thread with a conclusion, only to have my pc crash, twice...

Long story short: @williammartin convinced me and funnily enough it seems like I'm not the only one 🤝 

I'll get working on it soon and re-request review. Both of you or only one?